### PR TITLE
Feature/enhancements to student loans

### PR DESCRIPTION
**Pull request by crossland92**
Expanded functionality to include Student Loan plan 1 and 2 (instead of just 2), and allow it to be passed in via the command line, so people can select. 
Can expand to 3/4/5 in future (probably with a more efficient switch statement-esque function for calculating which is which.

### CODE REVIEW

File: uk_tax_23-24.py [hunk 1]
```diff
@@ -102,10 +116,11 @@ def calculate_taxes(incomes, pension_contrib_percent=0, voluntary_pension_contri
     student_loan_repayments = np.zeros_like(incomes)
 
     if include_student_loan:
-        student_loan_repayments = student_loan_repayment_plan_2(incomes)
+        student_loan_repayments = student_loan_repayment_plan(incomes,student_loan_plan_number)
 
     combined_taxes = income_taxes + national_insurances + student_loan_repayments
-    take_home_amounts = incomes - combined_taxes
+    # Need to account for any pension_contributions taken off here, if we've passed in a value
+    take_home_amounts = incomes_after_pension - combined_taxes
```
Comment (crossland92) [inline uk_tax_23-24.py:123]: Already merged into main just hadn't resynced

File: uk_tax_23-24.py [hunk 2]
```diff
@@ -24,6 +24,9 @@
 ni_rate_2 = 0.02
 
 # Student loan repayments
+student_loan_plan_1_threshold = 22015
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:27]: Maybe it makes sense to put the thresholds for all student loans in a dict? i.e.
```
student_loan_thresholds = {
  "Plan 1": 22015,
  "Plan 2": 27295,
  "Plan 4": 27660,
  "Plan 5": 25000,
  "Postgraduate": 21000
}
```

File: uk_tax_23-24.py [hunk 3]
```diff
@@ -80,16 +83,27 @@ def national_insurance(incomes):
...
-def student_loan_repayment_plan_2(incomes):
-    repayments = np.where(
-        incomes > student_loan_plan_2_threshold,
-        (incomes - student_loan_plan_2_threshold) * student_loan_plan_2_rate,
-        0,
-    )
-
+# Can expand for further student loan repayment plans (3/4/5) in future
+def student_loan_repayment_plan(incomes, student_loan_plan_number=2):
+    if student_loan_plan_number == 1:
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:88]: With the dictionary, we can avoid the nested ifs/elseifs:
```
def student_loan_repayment_plan(incomes, student_loan_plan="Plan 2"):

  student_loan_threshold = student_loan_thresholds[student_loan_plan]

  if student_loan_plan == "Postgraduate":
    student_loan_rate = 6
  else:
    student_loan_rate = 9

  repayments = np.where(
    incomes > student_loan_threshold,
      (incomes - student_loan_threshold) * (student_loan_rate / 100),
      0,
  )

  return repayments
```

I would also like to avoid printing anything in the console in this function, the top level function does that already. This function will be used to calculate the repayments for all salaries in the range and we don't want that printing each time.

File: uk_tax_23-24.py [hunk 4]
```diff
@@ -102,10 +116,11 @@ def calculate_taxes(incomes, pension_contrib_percent=0, voluntary_pension_contri
     student_loan_repayments = np.zeros_like(incomes)
 
     if include_student_loan:
-        student_loan_repayments = student_loan_repayment_plan_2(incomes)
+        student_loan_repayments = student_loan_repayment_plan(incomes,student_loan_plan_number)
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:119]: Missing space before the comma, sorry for being picky!

File: uk_tax_23-24.py [hunk 5]
```diff
@@ -80,16 +83,27 @@ def national_insurance(incomes):
...
+            0,
+        )
+    else:
+        print(f"We don't account for Student Loan plan number {student_loan_plan_number} yet - repayments will be 0")
+        repayments = 0
+    print(f"Student loan plan number is {student_loan_plan_number} and total repayments are \033[1;34m£{repayments}\033[0m")
     return repayments
 
-def calculate_taxes(incomes, pension_contrib_percent=0, voluntary_pension_contrib=0, include_student_loan=False):
+def calculate_taxes(incomes, pension_contrib_percent=0, student_loan_plan_number=2, voluntary_pension_contrib=0, include_student_loan=False):
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:106]: I would prefer if the Repayment Plan is a string ("Plan 2", "Postgraduate"), as it is more verbose and gives better support for the Postgrad Loan Repayments.

File: uk_tax_23-24.py [hunk 6]
```diff
@@ -195,14 +210,14 @@ def print_tax_breakdown(gross_income, pension_contrib_percent=0, voluntary_pensi
...
-            print(f"\033[1;34m Student Loan Repayment Plan 2:\033[0m £{student_loan_repayments[0]:.2f}")
+            print(f"\033[1;34m Student Loan Repayment Plan {student_loan_plan_number}:\033[0m £{student_loan_repayments[0]:.2f}")
 
         print(f"\033[1;34m Combined Taxes:\033[0m £{combined_taxes[0]:.2f}")
         print(f"\033[1;34m Take-home Amount:\033[0m £{take_home_amounts[0]:.2f}")
 
-def calculate_tax_savings(incomes, pension_contrib_percent, voluntary_pension_contrib):
-    taxes_no_contributions = calculate_taxes(incomes, pension_contrib_percent=0, voluntary_pension_contrib=0)
-    taxes_with_contributions = calculate_taxes(incomes, pension_contrib_percent, voluntary_pension_contrib)
+def calculate_tax_savings(incomes, pension_contrib_percent, voluntary_pension_contrib,student_loan_plan_number=2):
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:218]: Another comma missing here

File: uk_tax_23-24.py [hunk 7]
```diff
@@ -270,19 +285,35 @@ def plot_tax_savings_3d(salary_top_range=salary_top_range, max_voluntary_contrib
...
+    elif len(sys.argv) == 2: # Only 1 argument passed in (salary) - (student loan still defaults to 2 anyway and shows as 2 outputs, as per the original function)
         try:
             gross_income = float(sys.argv[1])
             if gross_income >= 0:
                 # Provide a breakdown for some specific gross income
                 print_tax_breakdown(gross_income)
                 plot_tax_savings_vs_pension_contributions(gross_income)
             else:
-                raise ValueError("Gross income must be a non-negative number.")
+                raise ValueError("Gross income must be a non-negative number")
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:311]: I liked it more with the dot, why did you remove it? lol

File: uk_tax_23-24.py [hunk 8]
```diff
@@ -270,19 +285,35 @@ def plot_tax_savings_3d(salary_top_range=salary_top_range, max_voluntary_contrib
         plt.show()
 
 if __name__ == "__main__":
-    if len(sys.argv) == 2:
+    # Expecting 1 or 2 arguments - 1 for salary (e.g 120000) and 1 for Student Loan Plan Number (1 or 2 currently)
+    if len(sys.argv) == 3: # All 2 arguments passed in (salary and student loan number)
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:289]: Could you also please update the README file to reflect the changes?

File: uk_tax_23-24.py [hunk 9]
```diff
@@ -270,19 +285,35 @@ def plot_tax_savings_3d(salary_top_range=salary_top_range, max_voluntary_contrib
...
+        try:
+            gross_income = float(sys.argv[1])
+            student_loan_plan_number = int(sys.argv[2])
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:292]: This can be a string (e.g. "Plan 2")

File: uk_tax_23-24.py [hunk 10]
```diff
@@ -270,19 +285,35 @@ def plot_tax_savings_3d(salary_top_range=salary_top_range, max_voluntary_contrib
...
+            if gross_income >= 0 and student_loan_plan_number >= 1:
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:293]: You can check if student_loan_plan is in student_loan_thresholds dictionary keys.

File: uk_tax_23-24.py [hunk 11]
```diff
@@ -270,19 +285,35 @@ def plot_tax_savings_3d(salary_top_range=salary_top_range, max_voluntary_contrib
...
+                # Provide a breakdown for some specific gross income
+                print(f"Student loan number passed is: {student_loan_plan_number}")
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:295]: This should move to the print_tax_breakdown function.

File: uk_tax_23-24.py [hunk 12]
```diff
@@ -270,19 +285,35 @@ def plot_tax_savings_3d(salary_top_range=salary_top_range, max_voluntary_contrib
...
                 print_tax_breakdown(gross_income)
                 plot_tax_savings_vs_pension_contributions(gross_income)
             else:
-                raise ValueError("Gross income must be a non-negative number.")
+                raise ValueError("Gross income must be a non-negative number")
         except ValueError as e:
             print(f"Invalid argument. Expected a number, got {sys.argv[1]}")
             print(str(e))
     else:
+        print(f"No args passed so plotting graphs only")
```
Comment (wozniakpawel) [inline uk_tax_23-24.py:316]: Unnecessary, I'd remove this line.

**Review by wozniakpawel — CHANGES_REQUESTED** [review]
Apologies for being a little bit harsh on the review, I'd like to keep the code nice and clean. Happy to merge with the requested changes!

**Comment by crossland92** [comment]
@wozniakpawel can see https://github.com/wozniakpawel/cooltaxtool/blob/master/src/utils/ incorporates a lot of these recommended changes into the delightful new website version, with plenty of other PRs layering onto there too. Due to that project likely being the one supported further, I'll leave these changes and see if anything pops up for me on the new repo and raise a PR there in future. Feel free to close this PR off
